server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -351,8 +358,70 @@ private <C extends Collector, T> T search(Weight weight, CollectorManager<C, T>
         }
     }
 
+    // Timestamp fields that may be correlated with the ingestion order, so that some segments are likely to be completely disjoint from the
+    // query range.
+    private static final Set<String> ECS_TIMESTAMP_FIELDS = Set.of("@timestamp", "event.ingested");
+
+    private static PointRangeQuery extractTimestampRange(Query query) {
+        // Remove query wrappers
+        while (true) {
+            if (query instanceof IndexOrDocValuesQuery idvQuery) {
+                query = idvQuery.getIndexQuery();
+            } else if (query instanceof ConstantScoreQuery csq) {
+                query = csq.getQuery();
+            } else if (query instanceof BoostQuery bq) {
+                query = bq.getQuery();
+            } else if (query instanceof IndexSortSortedNumericDocValuesRangeQuery issndvrq) {
+                query = issndvrq.getFallbackQuery();
+            } else {
+                break;
+            }
+        }
+
+        if (query instanceof PointRangeQuery pointRangeQuery) {
+            if (ECS_TIMESTAMP_FIELDS.contains(pointRangeQuery.getField())) {
+                return pointRangeQuery;
+            } else {
+                return null;
+            }
+        } else if (query instanceof BooleanQuery bq) {
+            for (BooleanClause clause : bq.clauses()) {
+                if (clause.isRequired()) {
+                    PointRangeQuery timestampRange = extractTimestampRange(clause.getQuery());
+                    if (timestampRange != null) {
+                        // In case multiple timestamp ranges are present, the first one that is found wins. This is fine as we expect
+                        // queries to have a single range on a timestamp field in general.
+                        return timestampRange;
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
     @Override
     public void search(List<LeafReaderContext> leaves, Weight weight, Collector collector) throws IOException {
+        // Alerting use-cases may run queries like the one below on a narrow time range:
+        // { "query": { "bool": { "must": <expensive query>, "filter": { "range": { "@timestamp": { "gte": "now-5m" } } } } } }
+        // This may cause performance issues when <expensive query> has a costly Weight#scorer implementation, since Lucene doesn't have a
+        // deterministic order for calling Weight#scorer, so it may call Weight#scorer on the expensive query before noticing that the
+        // segment doesn't even intersect with the time range. To prevent this, we proactively exclude segments that don't intersect with
+        // the time range. This can be seen as a can_match phase, but at the segment level.
```

Comment:
> This can be seen as a can_match phase, but at the segment level.

It really seems like this would be powerful not just for these time stamp ranges. But any query that we can determine if a segment is worth exploring in constant time. 

Are PointRangeQuery & PointValues the only way to have a constant time "segment can_match" ?

Reply:
No, there are other queries, e.g. `exists` queries could also skip segments if the field can't be found in field infos. The thing that is unique with these queries is that they target a field that correlates with the index order. So by design, they will tend to fully match some segments, be disjoint with some other segments, and only partially match one or two segments.

Reply:
> The thing that is unique with these queries is that they target a field that correlates with the index order.

Ah, so the implication is that checking other fields & queries just isn't worth it. 

I think this partial hack is ok, but it would be really nice if we could make this more general. 

e.g. with `exists` and other relevant queries that can make these sorts of constant time decisions. 

We also know with what fields the index is sorted, so we could expand the field names we care about (to the two proposed and any field name that the index is sorted by).

Reply:
Thanks for pushing on this... I think there is a better way of doing this indeed, I gave it a try at https://github.com/apache/lucene/pull/13033.